makeappointments helper

adibaejaz opened the pull request:
Prum 20ae_makeappointments
Helper for managing appointments in regolith

Returns members with a gap in their appointments
Returns members supported on an outdated grant
Returns members supported on a depleted grant
Returns overspent and underspent grants and a burn vs time plot
Allows user to make pseudo-appointments and pseudo-persons for the coming 4-5 months so that (1)-(4) are satisfied for the coming 1-2 years

sbillinge commented on regolith/helpers/makeappointmentshelper.py:
```diff
@@ -0,0 +1,134 @@
+    def sout(self):
+        rc = self.rc
+        gaps = []
+        outdated = []
```
to save space why not ``outdated, depleted, underspent, overspent = [], [], [], []``

adibaejaz replied:
done - sorry, Java habits are showing!

sbillinge replied:
don't apologise.  Nothing bad here, but we review to try and improve habits and make code more uniform.

adibaejaz replied:
yes, thank you :)

sbillinge commented on regolith/helpers/makeappointmentshelper.py:
```diff
@@ -0,0 +1,134 @@
+
+        for person in self.gtx["people"]:
+            appts = person.get("appointments")
+            if not is_fully_appointed(appts, begin_date=rc.begin_date, end_date=rc.end_date)[0]:
```
shouldn't this be ``begin_date=begin_date`` instead of ``begin_date=rc.begin_date`` (and so on)

adibaejaz replied:
done

sbillinge commented:
I am afraid this needs a rebase

sbillinge commented on regolith/helpers/makeappointmentshelper.py:
```diff
@@ -0,0 +1,126 @@
+def subparser(subpi):
+
+    subpi.add_argument("-v", "--verbose", action="store_true", help='increase verbosity of output')
+    #subpi.add_argument("-g", "--gap", action="store_true",
```
please clean

adibaejaz replied:
done

sbillinge commented on regolith/helpers/makeappointmentshelper.py:
```diff
@@ -0,0 +1,126 @@
+    #                   help='get group members supported on a depleted grant')
+    subpi.add_argument("-b", "--begin_date", help='begin date of interval to check appointment, to specify')
+    subpi.add_argument("-e", "--end_date", help='end date of interval to check appointments')
+    subpi.add_argument("-o", "--overspent", action="store_true", help='return grants that are overspent by more' 
```
I think I want this info without setting any flag.  Same for underspent

adibaejaz replied:
done

adibaejaz commented:
@sbillinge rebased!

adibaejaz commented on regolith/helpers/makeappointmentshelper.py:
```diff
@@ -0,0 +1,132 @@
+        for g in self.gtx["grants"]:
+            end_date = get_dates(g)['end_date']
+            if get_grant_amount(g, self.gtx["people"], end_date, end_date) > 100:
+                # what is the required 'underspend' amount?
```
this is just a crude inequality for the time being! what threshold for `student_months`, `postdoc_months` or `ss_months` should be used here?

sbillinge replied:
I took a different approach which was to plot the burn, but I like this functionality too.   Are the "units" student/postdoc months?  In that case probably an over or under of 3 months or more is about right.

adibaejaz replied:
thank you! I can also work on plotting the burn – i think that would look like extracting the values from this list and mapping time vs sum of ss_months, student_months and postdoc_months. so I should check if `student_months + postdoc_months < 3` on the `end_date` of the grant, correct? do we not factor in `ss_months`?

sbillinge replied:
yes, that's right, though check that the sum at the end is less than 1, not 3.  We should also put in ss months.   The burn plot looks funny because it has vertical steps on the anniversary of each grant as n student_monthe etc. are added in each budget year, and then slopes down as the appointments eat it up one day at a time.

adibaejaz replied:
done!

adibaejaz commented on regolith/helpers/makeappointmentshelper.py:
```diff
@@ -0,0 +1,132 @@
+            if get_grant_amount(g, self.gtx["people"], end_date, end_date) > 100:
+                # what is the required 'underspend' amount?
+                underspent.append(g.get('_id'))
+            elif get_grant_amount(g, self.gtx["people"], end_date, end_date) < 100:
```
same as above

sbillinge commented on regolith/helpers/makeappointmentshelper.py:
```diff
@@ -106,11 +107,29 @@ def sout(self):
+        if underspent:
+            print("Underspent grants:")
+            for g in underspent:
+                print(g)
```
I think it would be helpful to have the amount of the over/underspend (in student/pd - months)

adibaejaz replied:
done

sbillinge commented:
looks like there is a conflict and a CI problem.

sbillinge commented on tests/test_helpers.py:
```diff
@@ -85,6 +85,9 @@
      "above listed projectum ids that you would like to update.\n"
      ),
 ]
+    (["helper", "makeappointments"], "People with gaps in their appointments:\nAnthony Scopatz\nnot fully appointed on 2020-06-15\n"
```
this test seems to be not testing much of the intended functionality

adibaejaz replied:
Yes sorry about that Professor, unfortunately I am not able to push a more comprehensive version of the tests since the `people` schema only has one person with an `appointments` field, and their one appointment is on a grant that is not in the `EXEMPLARS` for grants, so this limits testing ground significantly. I am doing local schema edits to debug and test, but have not pushed these. Would you be open to me making changes to the schema as well, to allow for more testing?

sbillinge replied:
yes, you are supposed to modify the exemplars in the schema if you need more tests!    This may (or may not) affect other builders and we will have to deal with that, but better tests are always better than less good tests, and worth the effort in the long run.

adibaejaz replied:
working on this now that `grant_burn` has been merged!

sbillinge replied:
by far the best is to get the `grant_amount_function` functionsmerged then rebase onto updated master.    But if that isn't possible you can merge it in.   Maybe google around on this.  It might be better to squash when you merge and then cherry pick when we merge it into master to avoid that merge, but I am not sure.

adibaejaz replied:
great, thank you Professor! I will also do some stackoverflow-ing on this

adibaejaz commented on regolith/helpers/makeappointmentshelper.py:
```diff
@@ -0,0 +1,152 @@
+        all_appts = collect_appts(self.gtx["people"])
+
+
+        for p in self.gtx["people"]:
```
done

adibaejaz commented on tests/test_helpers.py:
```diff
@@ -159,6 +159,17 @@
+    (["helper", "makeappointments", "2019-08-31", "2019-10-31",],
+     "appointment gap for scopatz from 2019-08-31 to 2019-10-30\n"
+     "appointments on outdated grants:\n"
+     "person: scopatz, appointment: s20, grant: SymPy-1.1, from 2019-08-31 until 2020-05-15\n"
```
done

adibaejaz commented on tests/test_helpers.py:
```diff
@@ -159,6 +159,17 @@
      "Please rerun the helper specifying the complete ID.\n"
      ),
+    (["helper", "makeappointments", "2019-08-31", "2019-10-31",],
+     "appointment gap for scopatz from 2019-08-31 to 2019-10-30\n"
```
done

adibaejaz commented on tests/test_helpers.py:
```diff
@@ -159,6 +159,17 @@
+     "appointments on outdated grants:\n"
+     "person: scopatz, appointment: s20, grant: SymPy-1.1, from 2019-08-31 until 2020-05-15\n"
+     "appointments on depleted grants:\n"
+     "person: scopatz, appointment: f19, grant: dmref15, from 2019-09-01 until 2019-10-31\n"
```
done

adibaejaz commented on tests/test_helpers.py:
```diff
@@ -159,6 +159,17 @@
+     "appointments on depleted grants:\n"
+     "person: scopatz, appointment: f19, grant: dmref15, from 2019-09-01 until 2019-10-31\n"
+     "underspent grants:\n"
+     "2030-12-31: grant: SymPy-1.1, underspend amount: 8.0 months\n"
```
done

adibaejaz commented:
@sbillinge thank you Professor! I have:
- removed `begin_date` and `end_date` from the args altogether and defaulted them to the earliest/latest appointment dates for a specific person since this makes most sense for checking outdated/out-of-money appointments. I have not made them optional arguments either because I'm not really sure there is a use case for that, please let me know if there is and I will work on allowing optional dates
- added a blacklist of grants as was in the script you emailed me  and a schema exemplar for this just to make sure my code handles it correctly
and incorporated the other edits you suggested!

I have two more questions about what happens going forward:

- currently, the code only checks for overspent/underspent grants but does not return a plot of burn vs time. would you like me to include that functionality as well?
- in my recent commits, you will notice a huge chunk of code to tackle the 'simulation' functionality, it is very rough and rudimentary as I was just brainstorming ways of implementing it and hoping you could give me some feedback! however, if you need the previous checking functionalities asap, then perhaps i could move that to a new prum? please let me know!

adibaejaz commented:
@sbillinge I have removed the simulation code and added the plots!

adibaejaz commented:
@sbillinge thank you Professor! I noticed the incorrect rebase and have fixed this, so it is passing CI now, sorry about that! I have also added a news.
 are the new plots, burn for individual grants is plotted only for the lifetime of the grant so there is more clarity 
![image](https://user-images.githubusercontent.com/64800949/86540212-98153800-bf14-11ea-9f9e-3b5ca0448cfd.png)
![image](https://user-images.githubusercontent.com/64800949/86540218-9c415580-bf14-11ea-8e4f-f4742fcb46d5.png)
![image](https://user-images.githubusercontent.com/64800949/86540221-9fd4dc80-bf14-11ea-85a2-e72c24e2bb37.png)
![image](https://user-images.githubusercontent.com/64800949/86540227-a5322700-bf14-11ea-9ce4-a873e42dbcac.png)
The cumulative burn plot is the same as before.
Please let me know if you prefer these or the plots above! After this, I will remove WIP and request a merge!

sbillinge commented on regolith/helpers/makeappointmentshelper.py:
```diff
@@ -0,0 +1,194 @@
+                    this_ss[counter] = grant_amts[counter].get('ss_days')
+                    cum_ss[x] += grant_amts[counter].get('ss_days')
+                    counter += 1
+            plt.plot_date(grant_dates, this_student, ls='-', marker="", label="student days")
```
it is probably a good idea to pull all the plotting logic out into a function.  It won't be a very reusable function so leave it here in this module but maybe outside the helper class.  I am not sure how easily it can be tested, but if you can test it, put the tests into ``test_helpers.py``.

This will make it easier to maintain and turn it on and off if needs be.

adibaejaz replied:
done adding the function! testing might take a bit more time, however. also, would you like me to have a 
`subpi.add_argument("-p", "--plot", action="store_true", help='plot grant burns')
`
in the subparser?

sbillinge commented:
yes, I like this adaptive-range plot....let's stick with this.   Nice!
Please see my comment about moving plotting code to a function.  Then we
can merge.

On Sun, Jul 5, 2020 at 3:11 PM Adiba Ejaz <notifications@github.com> wrote:

> @sbillinge <https://github.com/sbillinge> thank you Professor! I noticed
> the incorrect rebase and have fixed this, so it is passing CI now, sorry
> about that! I have also added a news.
> are the new plots, burn for individual grants is plotted only for the
> lifetime of the grant so there is more clarity
> [image: image]
> <https://user-images.githubusercontent.com/64800949/86540212-98153800-bf14-11ea-9f9e-3b5ca0448cfd.png>
> [image: image]
> <https://user-images.githubusercontent.com/64800949/86540218-9c415580-bf14-11ea-8e4f-f4742fcb46d5.png>
> [image: image]
> <https://user-images.githubusercontent.com/64800949/86540221-9fd4dc80-bf14-11ea-85a2-e72c24e2bb37.png>
> [image: image]
> <https://user-images.githubusercontent.com/64800949/86540227-a5322700-bf14-11ea-9ce4-a873e42dbcac.png>
> The cumulative burn plot is the same as before.
> Please let me know if you prefer these or the plots above! After this, I
> will remove WIP and request a merge!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/451#issuecomment-653927144>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUI3X4UGIDMVTSKL5DDR2DF5ZANCNFSM4OBBKKXQ>
> .
>


-- 
Professor Simon Billinge
Columbia University

sbillinge commented:
Ooh, that’s a yes on the CL arg for plotting!  Great idea.   Actually I
think by default I want plot so maybe —no-plot (doesn’t need a short-name)
to suppress plotting.   I don’t really need tests on the plots tbh, though
we are getting into trouble with coverage and codecov so maybe just put a
smoke test on there...have the function return something if it runs without
error and check for that thing, sthg like that.

On Sun, Jul 5, 2020 at 3:28 PM Adiba Ejaz <notifications@github.com> wrote:

> *@adibaejaz* commented on this pull request.
> ------------------------------
>
> In regolith/helpers/makeappointmentshelper.py
> <https://github.com/regro/regolith/pull/451#discussion_r449910515>:
>
> > +                    str(grt_end), grant.get('_id'), round(grant_amt/30.5, 2)))
> +            elif grant_amt < -30.5:
> +                overspent.append("    {}: grant: {}, overspend amount: {} months".format(
> +                    str(grt_end), grant.get('_id'), round(grant_amt/30.5, 2)))
> +            counter = 0
> +            for x in range (grants_timespan.days + 1):
> +                if is_current(grant, now=datearray[x]):
> +                    grant_dates.append(datearray[x])
> +                    this_student[counter] = grant_amts[counter].get('student_days')
> +                    cum_student[x] += grant_amts[counter].get('student_days')
> +                    this_pd[counter] = grant_amts[counter].get('postdoc_days')
> +                    cum_pd[x] += grant_amts[counter].get('postdoc_days')
> +                    this_ss[counter] = grant_amts[counter].get('ss_days')
> +                    cum_ss[x] += grant_amts[counter].get('ss_days')
> +                    counter += 1
> +            plt.plot_date(grant_dates, this_student, ls='-', marker="", label="student days")
>
> done adding the function! testing might take a bit more time, however.
> also, would you like me to have a
> subpi.add_argument("-p", "--plot", action="store_true", help='plot grant
> burns')
> in the subparser?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/451#discussion_r449910515>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUPDDY42LFUKSRZ4FBTR2DH6ZANCNFSM4OBBKKXQ>
> .
>

adibaejaz commented:
@sbillinge I make the `plotter` function return "plotting mode is on", and print this in the cumulative plot function call, so I have added another test to `test_helpers.py` that checks that the helper does nor print this message when `--no-plot` is passed into the CLI! also, wrt Travis, is its alright that it is not recognising `matplotlib`?

sbillinge commented:
all good.  You have to add ``matplotlib-base`` into the run.txt in
requirements to make the CI work.

On Sun, Jul 5, 2020 at 4:06 PM Adiba Ejaz <notifications@github.com> wrote:

> @sbillinge <https://github.com/sbillinge> I make the plotter function
> return "plotting mode is on", and print this in the cumulative plot
> function call, so I have added another test to test_helpers.py that
> checks that the helper does nor print this message when --no-plot is
> passed into the CLI! also, wrt Travis, is its alright that it is not
> recognising matplotlib?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/451#issuecomment-653932896>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUIF353EOUVDGAJFYGLR2DMKPANCNFSM4OBBKKXQ>
> .
>


-- 
Professor Simon Billinge
Columbia University

adibaejaz commented:
@sbillinge fixed - now passing travis!

sbillinge commented:
We need to remove begin and end dates from the grants schema.  But I need
some help with sthg to make it work.  It is easier to explain in person.
Can you grab a slot tomorrow and we can talk about it?

On Sun, Jul 5, 2020 at 7:39 PM Adiba Ejaz <notifications@github.com> wrote:

> @sbillinge <https://github.com/sbillinge> fixed - now passing travis!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/451#issuecomment-653953853>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUJYM5SIUNZYO3YMZDTR2EFKHANCNFSM4OBBKKXQ>
> .
>
-- 
Professor Simon Billinge
Columbia University

adibaejaz commented:
@sbillinge I have added the proposal dates retrieval functionality!

sbillinge commented on regolith/helpers/makeappointmentshelper.py:
```diff
@@ -102,13 +103,16 @@ def sout(self):
                     continue
                 grant = rc.client.find_one(rc.database, "grants", {"_id": appt.get("grant")})
+                prop = rc.client.find_one(rc.database, "proposals", {"_id": grant.get("proposal_id")})
+                grant_begin, grant_end = get_dates(prop)['begin_date'], get_dates(prop)['end_date']
```
I think this assumes that the begin and end date will be in the proposal, but we only want to get it from the proposal if it is not found in the grant.  Sometimes these dates may be overwritten in the grant and what is in the grant should take precendent.  So only go look for them if they are not there in grant.  This was the behavior that merge_collections had, so we may want to go back to using that if it is a well tested function (though right now it is maybe not working).  I don't want to hold this up in case that takes a long time though, I urgently need this helper....

adibaejaz replied:
done! haven't used `merge_collections` here. I tested it by removing/ overwriting begin and end dates in some grants compared to their referenced proposals - however, I had to re-enter the grant dates in the schema so that the grant lister test would pass

sbillinge commented:
Not sure if it was ready for merge, but I saw the WIP gone.  Make sure to request a merge in a comment if you want it merged.

adibaejaz commented:
@sbillinge it was ready for merge! sorry, I will keep that in mind for future prums
